serverless/cmd/feeder/impl/feeder.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// impl is a witness feeder implementation for the serverless log.
+package impl
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/client"
+	"github.com/google/trillian/merkle/rfc6962/hasher"
+	"golang.org/x/mod/sumdb/note"
+)
+
+// Witness describes the operations the feeder needs to interact with a witness.
+type Witness interface {
+	// SigVerifier returns a verifier which can check signatures from this witness.
+	SigVerifier() note.Verifier
```

Comment:
I'm having trouble parsing this comment :)

But for the multiple keys thing, I think I'd prefer to leave as is for now - otherwise for any set up where there are more than 1 log (which should be ~ all of them unless there's a _really_ good reason why that's ok) we'd be expecting log believers to store a ~ quadratic number of public keys...

Reply:
Happy for this to stay as-is for now.

The unaddressed, and poorly communicated point was that if this method ends up taking a logID too, then every method on here does. Which implies that logID ought to not be something passed into the Witness, but a property of it.

Even if this method doesn't take one, I think there is still a case for the logID being a property per-Witness and not global across all witnesses for the feeder. It seems inevitable that there will be a future where two logs in different ecosystems pick the same logID (because it's just a friendly name at the moment), and then a witness would not be able to witness both of them, even though there is no technical reason why not.

Again, we can cross this bridge later if you prefer.

Reply:
Ah, gotcha.

Yeah, this interface is like this purely because the witness API is currently like this - I agree the impls could store the logID themselves and "augment" the calls they get from the feeder with this logID before passing it on to the actual witness.  I'll pop a TODO to revisit for now.

Personally, I think the `logID` should be defined to be more like `SHA256(public_key)`, and the tooling can have a friendly `Name` too (possibly just the one from the `note.Signer`)

Reply:
+1 to this. My comments are based on the friendly name situation at the moment.